### PR TITLE
HERA-428 [LMS] adds setting up medals from admin board

### PR DESCRIPTION
**Pull request by Golub-Sergey**
**Description:** adds setting up medals from admin board

**Youtrack:** [link](https://youtrack.raccoongang.com/issue/HERA-428)

### CODE REVIEW

File: lms/djangoapps/hera/models.py [hunk 1]
```diff
@@ -342,3 +342,37 @@ def reset_user_onboarding(sender, instance, *args, **kwargs):
     cache.delete(
         CACHING_USER_ONBOARDING_TEMPLATE.format(user_id=instance.user_id, model_name=sender.__name__)
     )
+
+
+class Medal(models.Model):
```
Comment (sendr) [inline lms/djangoapps/hera/models.py]: rename it pleasse to `MedalsSettings`

File: lms/djangoapps/hera/models.py [hunk 2]
```diff
@@ -342,3 +342,37 @@ def reset_user_onboarding(sender, instance, *args, **kwargs):
...
+    dna_icon = models.ImageField(upload_to='hera/medals', verbose_name="DNA icon")
+    min = models.PositiveSmallIntegerField()
+    max = models.PositiveSmallIntegerField()
+
+    def __str__(self):
+        return self.title
+
+    @classmethod
+    def _instance(cls):
+        return cls.objects.all().order_by('min')
```
Comment (sendr) [inline lms/djangoapps/hera/models.py]: please do so:
class Meta:
     ordering = ['min']

File: lms/djangoapps/hera/models.py [hunk 3]
```diff
@@ -342,3 +342,37 @@ def reset_user_onboarding(sender, instance, *args, **kwargs):
...
+            return instance
+
+    @classmethod
+    def get_medal(cls, points):
+        instance = cls._instance()
+        medal_title = ''
+        medal_dna_icon = ''
+        if instance:
+            for medal in instance:
+                if medal.min <= points <= medal.max:
```
Comment (sendr) [inline lms/djangoapps/hera/models.py]: is it possible to you the django lookups `min__gte=something` - or something like that?

File: lms/djangoapps/hera/utils.py [hunk 4]
```diff
@@ -1,16 +1,11 @@
...
-def get_medal(points):
-    if 90 <= points <= 100:
-        return "platinum"
-    elif 80 <= points <= 89:
-        return "gold"
-    elif 70 <= points <= 79:
-        return "silver"
-    elif 60 <= points <= 69:
-        return "copper"
+try:
```
Comment (sendr) [inline lms/djangoapps/hera/utils.py]: is it possible to relocate the `hera` app to `common/djangoapps` directory?

File: lms/djangoapps/hera/models.py (hunk 1 above)
Comment (Golub-Sergey) [reply lms/djangoapps/hera/models.py]: Done

File: lms/djangoapps/hera/models.py (hunk 2 above)
Comment (Golub-Sergey) [reply lms/djangoapps/hera/models.py]: Done

File: lms/djangoapps/hera/models.py (hunk 3 above)
Comment (Golub-Sergey) [reply lms/djangoapps/hera/models.py]: Done

File: lms/djangoapps/hera/utils.py (hunk 4 above)
Comment (Golub-Sergey) [reply lms/djangoapps/hera/utils.py]: We have some hera lms templates in the edx-theme, do we need to relocate edx-theme templates for correct hera templates mapping?

File: lms/djangoapps/hera/admin.py [hunk 5]
```diff
@@ -120,8 +120,39 @@ def has_add_permission(self, request):
...
+    def clean(self):
+        cleaned_data = super(MedalFormAdmin, self).clean()
+        min = cleaned_data.get("min")
+        max = cleaned_data.get("max")
+        if min > max:
+            raise forms.ValidationError("min value should be less then max")
+
+    def clean_min(self):
+        min = self.cleaned_data['min']
+        if not min <= 99:
```
Comment (sendr) [inline lms/djangoapps/hera/admin.py]:
```suggestion
        if min >= 99:
```

File: lms/djangoapps/hera/admin.py [hunk 6]
```diff
@@ -120,8 +120,39 @@ def has_add_permission(self, request):
...
+            raise forms.ValidationError("max value should be less then or equal to 99")
+        return min
+
+    def clean_max(self):
+        max = self.cleaned_data['max']
+        if not max <= 100:
```
Comment (sendr) [inline lms/djangoapps/hera/admin.py]:
```suggestion
        if max >= 100:
```

File: lms/djangoapps/hera/models.py [hunk 7]
```diff
@@ -295,9 +295,9 @@ def delete(self, *args, **kwargs):
 
 
 class Mascot(models.Model):
-    user_dashboard = models.ImageField(upload_to='hera', null=True, blank=True)
-    user_dashboard_modal = models.ImageField(upload_to='hera', null=True, blank=True)
-    onboarding = models.ImageField(upload_to='hera', null=True, blank=True)
+    user_dashboard = models.ImageField(upload_to='hera/mascots', null=True, blank=True)
```
Comment (sendr) [inline lms/djangoapps/hera/models.py]: is it worth making some constant for `upload_to` attribute not to repeat this string three times?

File: lms/djangoapps/hera/models.py [hunk 8]
```diff
@@ -363,3 +363,38 @@ def reset_user_onboarding(sender, instance, *args, **kwargs):
...
+
+    class Meta:
+        ordering = ['min']
+        verbose_name_plural = "Medals settings"
+
+    def __str__(self):
+        return self.title
+
+    @classmethod
+    def _instance(cls):
```
Comment (sendr) [inline lms/djangoapps/hera/models.py]: why do you name it `_instance` if it's the Queryset?

File: lms/djangoapps/hera/utils.py (hunk 4 above)
Comment (sendr) [reply lms/djangoapps/hera/utils.py]: if it's needed and will bring results and won't take a long time - yes).

File: lms/djangoapps/hera/admin.py [hunk 9]
```diff
@@ -120,8 +120,39 @@ def has_add_permission(self, request):
...
+
+    def clean_max(self):
+        max = self.cleaned_data['max']
+        if max > 100:
+            raise forms.ValidationError("max value should be less then or equal to 100")
+        return max
+
+    class Meta:
+        model = MedalsSettings
+        fields = ('title', 'image', 'dna_icon', 'min', 'max',)
```
[the remark this answers is not in the document]
Comment (Golub-Sergey) [reply lms/djangoapps/hera/admin.py]: Done

File: lms/djangoapps/hera/utils.py [hunk 10]
```diff
@@ -44,7 +36,8 @@ def get_lesson_summary_xblock_context(user, course_key, current_unit):
                         points = 0
                     break
 
-    return {'lesson_title': lesson_title, 'points': points, 'medal': get_medal(points)}
+    medal, _ = MedalsSettings.get_medal(points)
+    return {'lesson_title': lesson_title, 'points': points, 'medal': medal}
```
Comment (sendr) [inline lms/djangoapps/hera/utils.py]:
```suggestion
    return {
        'lesson_title': lesson_title,
        'points': points,
        'medal': medal
    }
```

File: lms/djangoapps/hera/fragments.py [hunk 11]
```diff
@@ -290,7 +289,10 @@ def render_to_fragment(self, request, course_id=None, page_context=None, **kwarg
                             points = int(subsection.all_total.earned / subsection.all_total.possible * 100)
                         except ZeroDivisionError:
                             points = 0
-                        completed_subsections[subsection.location.block_id].update({'points': points, 'medal': get_medal(points)})
+                        medal, dna_icon = MedalsSettings.get_medal(points)
+                        completed_subsections[subsection.location.block_id].update({
+                            'points': points, 'medal': medal, 'medal_dna_icon': dna_icon
```
[the remark this answers is not in the document]
Comment (Golub-Sergey) [reply lms/djangoapps/hera/fragments.py]: done

File: lms/djangoapps/hera/models.py [hunk 12]
```diff
@@ -295,9 +295,11 @@ def delete(self, *args, **kwargs):
 
 
 class Mascot(models.Model):
-    user_dashboard = models.ImageField(upload_to='hera', null=True, blank=True)
-    user_dashboard_modal = models.ImageField(upload_to='hera', null=True, blank=True)
-    onboarding = models.ImageField(upload_to='hera', null=True, blank=True)
+    MASCOT_IMAGE_PATH = 'hera/medals'
```
[the remark this answers is not in the document]
Comment (Golub-Sergey) [reply lms/djangoapps/hera/models.py]: done

File: lms/djangoapps/hera/utils.py (hunk 10 above)
Comment (Golub-Sergey) [reply lms/djangoapps/hera/utils.py]: done

**Pull request merged**